components/analytics/true_up_review.tsx

```diff
@@ -0,0 +1,217 @@
+import {
+    isCurrentLicenseCloud,
+} from 'mattermost-redux/selectors/entities/cloud';
+
+import {
+    getSelfHostedErrors,
+    getTrueUpReviewProfile as trueUpReviewProfileSelector,
+    getTrueUpReviewStatus as trueUpReviewStatusSelector,
+} from 'mattermost-redux/selectors/entities/hosted_customer';
+
+import useCanSelfHostedSignup from 'components/common/hooks/useCanSelfHostedSignup';
+import CheckMarkSvg from 'components/widgets/icons/check_mark_icon';
+
+import './true_up_review.scss';
+import {GlobalState} from '@mattermost/types/store';
+import WarningIcon from 'components/widgets/icons/fa_warning_icon';
+import {isCurrentUserSystemAdmin} from 'mattermost-redux/selectors/entities/users';
+import {getLicense} from 'mattermost-redux/selectors/entities/general';
+
+const TrueUpReview: React.FC = () => {
+    const dispatch = useDispatch();
+    const isCloud = useSelector(isCurrentLicenseCloud);
+    const isAirGapped = !useCanSelfHostedSignup();
+    const reviewProfile = useSelector(trueUpReviewProfileSelector);
+    const reviewStatus = useSelector(trueUpReviewStatusSelector);
+    const isSystemAdmin = useSelector(isCurrentUserSystemAdmin);
+    const license = useSelector(getLicense);
+    const [submitted, setSubmitted] = useState(false);
+    const trueUpReviewError = useSelector((state: GlobalState) => {
+        const errors = getSelfHostedErrors(state);
+        return Boolean(errors.trueUpReview);
+    });
+
+    useEffect(() => {
+        dispatch(getTrueUpReviewStatus());
+    }, [dispatch, getTrueUpReviewStatus]);
+
+    // Download the review profile as a base64 encoded json file when the review request is submitted.
+    useEffect(() => {
+        if (submitted && isAirGapped) {
+            // Create the bundle as a blob containing base64 encoded json data and assign it to a link element.
+            const blob = new Blob([reviewProfile.content], {type: 'application/text'});
+            const href = URL.createObjectURL(blob);
+
+            const link = document.createElement('a');
+            const date = moment().format('MM-DD-YYYY');
+            link.href = href;
+            link.download = `True Up-${license.Id}-${date}.txt`;
+            document.body.appendChild(link);
+            link.click();
+
+            // Remove link and revoke object url to avoid memory leaks.
+            document.body.removeChild(link);
+            URL.revokeObjectURL(href);
+
+            dispatch(getTrueUpReviewStatus());
+        }
+    }, [isAirGapped, submitted, reviewProfile]);
+
+    const formattedDueDate = (): string => {
+        if (!reviewStatus?.due_date) {
+            return '';
+        }
+
+        // Convert from milliseconds
+        const date = new Date(reviewStatus.due_date);
+        return moment(date).format('MMMM DD, YYYY');
+    };
+
+    const handleSubmitReview = () => {
+        dispatch(getTrueUpReviewBundle());
+        dispatch(getTrueUpReviewStatus());
+        setSubmitted(true);
```

Comment:
Please let me know if I'm wrong. Does' getTrueUpReviewStatus' depend on `getTrueUpReviewBundle` and `setSubmitted ` depending on the previous requests? If the answer is yes, we have two bugs:
- The first bug is calling the second request even if the first didn't complete or failed.
- The second bug is updating the `setSubmitted` status without waiting and checking the request status. The reviewBundle request can fail and try to download a file that doesn't exist. 

If the requests are not dependent, the second bug still applies.  You can use async, [an example of this](https://github.com/mattermost/mattermost-webapp/blob/6f4586eef4585d02b1e0515955bc00dd7336f74a/components/payment_form/gather_intent/useGatherIntent.ts#L40)

Reply:
`getTrueUpReviewStatus` does not rely on `getTrueUpReviewBundle`, it's more the other way around.

If `getTrueUpReviewStatus` fails, an error state is shown and the user cannot request a review.
I think my intention here with calling `getTrueUpReviewStatus` was just to have the updated status (i.e. completed or not) present.

That example is super helpful though. The reason for the `submitted` state even existing is so that I can ensure I have the bundle before downloading it. I'll see if I can rework this section with this new info. Thanks :)